openstudiocore/src/model/ZonePropertyUserViewFactorsBySurfaceName.cpp

```diff
@@ -0,0 +1,505 @@
+
+    if (!_toSurface) {
+      LOG(Error, "Could not retrieve FromSurfaceName for extensible group " << group.groupIndex() << ".");
+    }
+    if (!_fromSurface) {
+      LOG(Error, "Could not retrieve ToSurfaceName for extensible group " << group.groupIndex() << ".");
+    }
+    if (!value) {
+      LOG(Error, "Could not retrieve ViewFactor for extensible group " << group.groupIndex() << ".");
+    }
+
+    result = ViewFactor(_toSurface.get(), _fromSurface.get(), value.get());
+
+    return result;
+
+  }
+
+  bool ZonePropertyUserViewFactorsBySurfaceName_Impl::addViewFactor(const ViewFactor& viewFactor) {
+    bool result = false;
+
+
+    // This the only place we need to cast, in order to check if the actual to/from stuff is part of our zone
+    // This lambda checks that the InternalMass/Surface/SubSurface is part of the ThermalZone referenced by this object
+    auto checkIfInThermalZone = [this](const ModelObject& modelObject) {
+      bool result = false;
+      boost::optional<Space> _space;
+
+      if (boost::optional<Surface> _fromSurfaceAsSurface = modelObject.optionalCast<Surface>()) {
+        _space = _fromSurfaceAsSurface->space();
+      } else if (boost::optional<SubSurface> _fromSurfaceAsSubSurface = modelObject.optionalCast<SubSurface>()) {
+        _space = _fromSurfaceAsSubSurface->space();
+      } else if (boost::optional<InternalMass> _fromSurfaceAsInternalMass = modelObject.optionalCast<InternalMass>()) {
+        _space = _fromSurfaceAsInternalMass->space();
+      } else {
+        // This can't happen, ViewFactor shouldn't accept anything else
+        OS_ASSERT(false);
+      }
+
+      if (_space) {
+        if (boost::optional<ThermalZone> _zone = _space->thermalZone()) {
+          if (_zone->handle() == this->thermalZone().handle()) {
+            result = true;
+          }
+        }
+      }
+      return result;
+    };
+
+    ModelObject fromSurface = viewFactor.fromSurface();
+    if (!checkIfInThermalZone(fromSurface)) {
+      LOG(Error, "Cannot add ViewFactor to " << briefDescription() << " because fromSurface=" << fromSurface.briefDescription()
+          << "'is not part of the ThermalZone.");
+      return false;
+    }
+
+    ModelObject toSurface = viewFactor.toSurface();
+    if (!checkIfInThermalZone(toSurface)) {
+      LOG(Error, "Cannot add ViewFactor to " << briefDescription() << " because toSurface=" << toSurface.briefDescription()
+          << "'is not part of the ThermalZone.");
+      return false;
+    }
+
+    // Check if viewFactor already exists
+    boost::optional<unsigned> _existingIndex = viewFactorIndex(viewFactor);
+    if (_existingIndex) {
+      boost::optional<ViewFactor> _viewFactor = getViewFactor(_existingIndex.get());
+      OS_ASSERT(_viewFactor);
+      LOG(Warn, "For " << briefDescription() << ", ViewFactor already exists, will be modified in place from " << _viewFactor.get()
```

Comment:
Do we want the view factor to be overwritten, or ignored (and the existing one is left untouched)?

@shorowit

Reply:
I'd overwrite it if it's not too much work.

Reply:
OK. That's how it's written currently.